ngsPETSc/utils/firedrake/adaptive_hierarchy.py

```diff
@@ -0,0 +1,331 @@
+"""
+This module contains the class for the AdaptiveMeshHierarchy and related helper functions
+"""
+from fractions import Fraction
+from collections import defaultdict
+import numpy as np
+from firedrake import (Function, Cofunction, Mesh, Submesh, RelabeledMesh,
+                       HierarchyBase, FunctionSpace, conditional, gt
+)
+from firedrake.mg.utils import set_level, get_level
+
+__all__ = ["AdaptiveMeshHierarchy"]
+
+
+class AdaptiveMeshHierarchy(HierarchyBase):
+    """
+    HierarchyBase for hierarchies of adaptively refined meshes
+    """
+    def __init__(self, mesh, refinements_per_level=1, nested=True):
+        self.meshes = tuple(mesh)
+        self._meshes = tuple(mesh)
+        self.submesh_hierarchies = []
+        self.coarse_to_fine_cells = None
+        self.fine_to_coarse_cells = None
```

Comment:
Can these be initialized here so that we don't have to do it in `add_mesh`?
```suggestion
        self.coarse_to_fine_cells = {}
        self.fine_to_coarse_cells = {}
```